plugin/trino-sqlserver/src/main/java/io/trino/plugin/sqlserver/SqlServerClient.java

```diff
@@ -441,6 +477,72 @@ public void setNull(PreparedStatement statement, int index)
         };
     }
 
+    public static ColumnMapping timestampColumnMapping(TimestampType timestampType)
+    {
+        if (timestampType.getPrecision() <= MAX_SHORT_PRECISION) {
+            // SQLServer JDBC driver does not support LocalDateTime so timestampColumnMapping() cannot be used here
+            return ColumnMapping.longMapping(
+                    timestampType,
+                    shortTimestampReadFunction(),
+                    shortTimestampWriteFunction());
+        }
+        return ColumnMapping.objectMapping(
+                timestampType,
+                longTimestampReadFunction(),
+                longTimestampWriteFunction());
+    }
+
+    private static LongReadFunction shortTimestampReadFunction()
+    {
+        return (resultSet, columnIndex) -> {
+            Timestamp timestamp = resultSet.getTimestamp(columnIndex, newUtcCalendar());
+            int nanoSeconds = timestamp.getNanos() % NANOSECONDS_PER_MILLISECOND;
+            return timestamp.getTime() * MICROSECONDS_PER_MILLISECOND + nanoSeconds / NANOSECONDS_PER_MICROSECOND;
+        };
+    }
+
+    private static LongWriteFunction shortTimestampWriteFunction()
+    {
+        return (statement, index, value) -> {
+            long epochSeconds = floorDiv(value, MICROSECONDS_PER_SECOND);
+            int microOfSecond = floorMod(value, MICROSECONDS_PER_SECOND);
+            int nanosOfSecond = (microOfSecond * NANOSECONDS_PER_MICROSECOND);
+
+            Timestamp timestamp = new Timestamp(epochSeconds * MILLISECONDS_PER_SECOND);
+            timestamp.setNanos(nanosOfSecond);
+            statement.setTimestamp(index, timestamp, newUtcCalendar());
+        };
+    }
+
+    private static ObjectReadFunction longTimestampReadFunction()
+    {
+        return ObjectReadFunction.of(
+                LongTimestamp.class,
+                (resultSet, columnIndex) -> {
+                    Timestamp timestamp = resultSet.getTimestamp(columnIndex, newUtcCalendar());
+                    int nanoSeconds = timestamp.getNanos() % NANOSECONDS_PER_MILLISECOND;
+                    long epochMicros = timestamp.getTime() * MICROSECONDS_PER_MILLISECOND + nanoSeconds / NANOSECONDS_PER_MICROSECOND;
+                    int picosOfMicro = (nanoSeconds % NANOSECONDS_PER_MICROSECOND) * PICOSECONDS_PER_NANOSECOND;
```

Comment:
should we verify that `picosOfMicro % 10^5 == 0` (as max precision is 7) ? @findepi not sure if we do that for other databases.

Reply:
for others connectors we're not doing it correct -- https://github.com/trinodb/trino/issues/6607

Reply:
Should I add a call to `verify()` here or round the value? This is only necessary for reads, not writes, right?

Reply:
I added `verify` calls to both short and long readers, please review again.